File: mashora-site/backend/app/routers/addons.py
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy import func, or_, select
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import selectinload

from app.database import get_db
from app.middleware.auth import get_current_user
from app.models import Addon, AddonReview, AddonVersion, Tenant, TenantAddon, User
from app.schemas.addon import (
    AddonDetail,
    AddonList,
    AddonResponse,
    AddonReviewCreate,
    AddonReviewResponse,
    AddonVersionResponse,
    InstallAddonRequest,
)

router = APIRouter(prefix="/addons", tags=["addons"])


@router.get("", response_model=AddonList)
async def browse_addons(
    q: str | None = Query(None),
    category: str | None = Query(None),
    page: int = Query(1, ge=1),
    per_page: int = Query(20, ge=1, le=100),
    sort_by: str = Query("download_count"),
    db: AsyncSession = Depends(get_db),
):
    query = select(Addon).where(Addon.status == "published")

    if q:
        query = query.where(
            or_(
                Addon.technical_name.ilike(f"%{q}%"),
                Addon.display_name.ilike(f"%{q}%"),
                Addon.summary.ilike(f"%{q}%"),
            )
        )
    if category:
        query = query.where(Addon.category == category)

    # Count total
    count_q = select(func.count()).select_from(query.subquery())
    total = (await db.execute(count_q)).scalar() or 0

    # Sort
    sort_col = getattr(Addon, sort_by, Addon.download_count)
    query = query.order_by(sort_col.desc())

    # Paginate
    query = query.offset((page - 1) * per_page).limit(per_page)
    result = await db.execute(query)
    addons = result.scalars().all()

    return AddonList(
        addons=[AddonResponse.model_validate(a) for a in addons],
        total=total,
        page=page,
        per_page=per_page,
    )


@router.get("/{technical_name}", response_model=AddonDetail)
async def get_addon(technical_name: str, db: AsyncSession = Depends(get_db)):
    result = await db.execute(
        select(Addon)
        .where(Addon.technical_name == technical_name)
        .options(selectinload(Addon.versions))
    )
    addon = result.scalar_one_or_none()
    if addon is None:
        raise HTTPException(status_code=404, detail="Addon not found")

    versions = [AddonVersionResponse.model_validate(v) for v in addon.versions]
    data = AddonResponse.model_validate(addon).model_dump()
    data["description"] = addon.description
    data["mashora_version_min"] = addon.mashora_version_min
    data["versions"] = versions
    return AddonDetail(**data)


@router.post("/{technical_name}/install")
async def install_addon(
    technical_name: str,
    body: InstallAddonRequest,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    # Find addon
    result = await db.execute(select(Addon).where(Addon.technical_name == technical_name))
    addon = result.scalar_one_or_none()
    if addon is None:
        raise HTTPException(status_code=404, detail="Addon not found")

    # Verify tenant belongs to user's org
    result = await db.execute(
        select(Tenant).where(Tenant.id == body.tenant_id, Tenant.org_id == current_user.org_id)
    )
    tenant = result.scalar_one_or_none()
    if tenant is None:
        raise HTTPException(status_code=403, detail="Tenant not found or not owned by you")

    # Get latest version if not specified
    version_id = body.version_id
    if version_id is None:
        result = await db.execute(
            select(AddonVersion)
            .where(AddonVersion.addon_id == addon.id)
            .order_by(AddonVersion.published_at.desc())
            .limit(1)
        )
        latest = result.scalar_one_or_none()
        if latest is None:
            raise HTTPException(status_code=400, detail="No versions available")
        version_id = latest.id

    # Check if already installed
    result = await db.execute(
        select(TenantAddon).where(
            TenantAddon.tenant_id == body.tenant_id, TenantAddon.addon_id == addon.id
        )
    )
    if result.scalar_one_or_none() is not None:
        raise HTTPException(status_code=409, detail="Addon already installed on this tenant")

    # Install
    ta = TenantAddon(tenant_id=body.tenant_id, addon_id=addon.id, addon_version_id=version_id)
    db.add(ta)
    addon.download_count += 1
    await db.flush()

    return {"message": "Addon installed successfully"}


@router.post("/{technical_name}/review", response_model=AddonReviewResponse)
async def review_addon(
    technical_name: str,
    body: AddonReviewCreate,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    result = await db.execute(select(Addon).where(Addon.technical_name == technical_name))
    addon = result.scalar_one_or_none()
    if addon is None:
        raise HTTPException(status_code=404, detail="Addon not found")

    # Upsert review
    result = await db.execute(
        select(AddonReview).where(
            AddonReview.addon_id == addon.id, AddonReview.user_id == current_user.id
        )
    )
    review = result.scalar_one_or_none()
    if review:
        review.rating = body.rating
        review.comment = body.comment
    else:
        review = AddonReview(
            addon_id=addon.id, user_id=current_user.id, rating=body.rating, comment=body.comment
        )
        db.add(review)

    await db.flush()

    # Recalculate average
    result = await db.execute(
        select(func.avg(AddonReview.rating), func.count()).where(
            AddonReview.addon_id == addon.id
        )
    )
    avg, count = result.one()
    addon.rating_avg = float(avg or 0)
    addon.rating_count = count
    await db.flush()

    return AddonReviewResponse(
        id=review.id,
        user_id=review.user_id,
        user_email=current_user.email,
        rating=review.rating,
        comment=review.comment,
        created_at=review.created_at,
    )


@router.get("/{technical_name}/reviews", response_model=list[AddonReviewResponse])
async def list_reviews(technical_name: str, db: AsyncSession = Depends(get_db)):
    result = await db.execute(select(Addon).where(Addon.technical_name == technical_name))
    addon = result.scalar_one_or_none()
    if addon is None:
        raise HTTPException(status_code=404, detail="Addon not found")

    result = await db.execute(
        select(AddonReview, User.email)
        .join(User, AddonReview.user_id == User.id)
        .where(AddonReview.addon_id == addon.id)
        .order_by(AddonReview.created_at.desc())
    )
    rows = result.all()
    return [
        AddonReviewResponse(
            id=r.AddonReview.id,
            user_id=r.AddonReview.user_id,
            user_email=r.email,
            rating=r.AddonReview.rating,
            comment=r.AddonReview.comment,
            created_at=r.AddonReview.created_at,
        )
        for r in rows
    ]

File: mashora-site/frontend/src/api/addons.ts
import client from './client'

export interface AddonResponse {
  id: string
  technical_name: string
  display_name: string
  summary: string
  author_name: string | null
  category: string
  version: string
  price_cents: number
  currency: string
  icon_url: string | null
  download_count: number
  rating_avg: number
  rating_count: number
  status: string
  created_at: string
}

export interface AddonVersionResponse {
  id: string
  version: string
  changelog: string
  file_size: number
  mashora_version_compat: string
  published_at: string
}

export interface AddonDetail extends AddonResponse {
  description: string
  mashora_version_min: string
  versions: AddonVersionResponse[]
}

export interface AddonReviewResponse {
  id: string
  user_id: string
  user_email: string | null
  rating: number
  comment: string
  created_at: string
}

export interface AddonList {
  addons: AddonResponse[]
  total: number
  page: number
  per_page: number
}

function normalizeAddon(data: any): AddonResponse {
  return {
    id: String(data.id),
    technical_name: data.technical_name ?? '',
    display_name: data.display_name ?? '',
    summary: data.summary ?? '',
    author_name: data.author_name ?? null,
    category: data.category ?? 'Other',
    version: data.version ?? '',
    price_cents: Number(data.price_cents ?? 0),
    currency: data.currency ?? 'USD',
    icon_url: data.icon_url ?? null,
    download_count: Number(data.download_count ?? 0),
    rating_avg: Number(data.rating_avg ?? 0),
    rating_count: Number(data.rating_count ?? 0),
    status: data.status ?? 'pending',
    created_at: data.created_at ?? new Date().toISOString(),
  }
}

function normalizeAddonVersion(data: any): AddonVersionResponse {
  return {
    id: String(data.id),
    version: data.version ?? '',
    changelog: data.changelog ?? '',
    file_size: Number(data.file_size ?? 0),
    mashora_version_compat: data.mashora_version_compat ?? '',
    published_at: data.published_at ?? new Date().toISOString(),
  }
}

function normalizeAddonReview(data: any): AddonReviewResponse {
  return {
    id: String(data.id),
    user_id: String(data.user_id),
    user_email: data.user_email ?? null,
    rating: Number(data.rating ?? 0),
    comment: data.comment ?? '',
    created_at: data.created_at ?? new Date().toISOString(),
  }
}

export async function browseAddons(params: {
  q?: string
  category?: string
  page?: number
  per_page?: number
  sort_by?: string
}): Promise<AddonList> {
  const res = await client.get('/addons', { params })
  return {
    addons: (res.data.addons ?? []).map(normalizeAddon),
    total: Number(res.data.total ?? 0),
    page: Number(res.data.page ?? 1),
    per_page: Number(res.data.per_page ?? params.per_page ?? 20),
  }
}

export async function getAddon(technicalName: string): Promise<AddonDetail> {
  const res = await client.get(`/addons/${technicalName}`)
  return {
    ...normalizeAddon(res.data),
    description: res.data.description ?? '',
    mashora_version_min: res.data.mashora_version_min ?? '19.0',
    versions: (res.data.versions ?? []).map(normalizeAddonVersion),
  }
}

export async function installAddon(
  technicalName: string,
  tenantId: string,
  versionId?: string
): Promise<void> {
  await client.post(`/addons/${technicalName}/install`, { tenant_id: tenantId, version_id: versionId })
}

export async function reviewAddon(
  technicalName: string,
  rating: number,
  comment?: string
): Promise<AddonReviewResponse> {
  const res = await client.post(`/addons/${technicalName}/review`, { rating, comment })
  return normalizeAddonReview(res.data)
}

export async function getAddonReviews(technicalName: string): Promise<AddonReviewResponse[]> {
  const res = await client.get(`/addons/${technicalName}/reviews`)
  return (res.data ?? []).map(normalizeAddonReview)
}

export async function submitAddon(data: {
  technical_name: string
  display_name: string
  summary: string
  description: string
  category: string
  price_cents?: number
}): Promise<AddonResponse> {
  const res = await client.post('/publisher/addons', data)
  return normalizeAddon(res.data)
}

export async function getPublisherAddons(): Promise<AddonResponse[]> {
  const res = await client.get('/publisher/addons')
  return (res.data ?? []).map(normalizeAddon)
}

export async function uploadVersion(
  technicalName: string,
  formData: FormData
): Promise<AddonVersionResponse> {
  const res = await client.post(`/publisher/addons/${technicalName}/versions`, formData, {
    headers: { 'Content-Type': 'multipart/form-data' },
  })
  return normalizeAddonVersion(res.data)
}
